packages/core/src/codewhispererChat/controllers/chat/messenger/messenger.ts

```diff
@@ -520,48 +584,92 @@ export class Messenger {
         )
     }
 
+    private sendReadAndListDirToolMessage(
+        toolUse: ToolUse,
+        session: ChatSession,
+        tabID: string,
+        triggerID: string,
+        messageIdToUpdate?: string
+    ) {
+        const contextList = toolUse.name === ToolType.ListDirectory ? session.readFolders : session.readFiles
+        const isFileRead = toolUse.name === ToolType.FsRead
+        const items = isFileRead ? session.readFiles : session.readFolders
+        const itemCount = items.length
+
+        const title =
+            itemCount < 1
+                ? 'Gathering context'
+                : isFileRead
+                  ? `${itemCount} file${itemCount > 1 ? 's' : ''} read`
+                  : `${itemCount} ${itemCount === 1 ? 'directory' : 'directories'} listed`
+
+        this.dispatcher.sendToolMessage(
+            new ToolMessage(
+                {
+                    message: '',
+                    messageType: 'answer-part',
+                    followUps: undefined,
+                    followUpsHeader: undefined,
+                    relatedSuggestions: undefined,
+                    triggerID,
+                    messageID: messageIdToUpdate ?? toolUse?.toolUseId ?? '',
+                    userIntent: undefined,
+                    codeBlockLanguage: undefined,
+                    contextList,
+                    canBeVoted: false,
+                    buttons: undefined,
+                    fullWidth: false,
+                    padding: false,
+                    codeBlockActions: undefined,
+                    title,
+                },
+                tabID
+            )
+        )
+    }
+
     public sendPartialToolLog(
         message: string,
         tabID: string,
         triggerID: string,
         toolUse: ToolUse | undefined,
+        session: ChatSession,
+        messageIdToUpdate: string | undefined,
         validation: CommandValidation,
         changeList?: Change[]
     ) {
+        // Handle read tool and list directory messages
+        if (toolUse?.name === ToolType.FsRead || toolUse?.name === ToolType.ListDirectory) {
+            return this.sendReadAndListDirToolMessage(toolUse, session, tabID, triggerID, messageIdToUpdate)
+        }
```

Comment:
nit: this is still confusing, I think the condition should not happen in this function and always call processToolMessage